src/librustc_codegen_llvm/builder.rs

```diff
@@ -494,9 +527,80 @@ impl Builder<'a, 'll, 'tcx> {
         }
     }
 
+    fn load_ref(
+        &mut self,
+        ptr: &PlaceRef<'tcx, &'ll Value>
+    ) -> OperandRef<'tcx, &'ll Value> {
+        debug!("PlaceRef::load: {:?}", ptr);
+
+        assert_eq!(ptr.llextra.is_some(), ptr.layout.is_unsized());
 
-    pub fn range_metadata(&self, load: &'ll Value, range: Range<u128>) {
-        if self.sess().target.target.arch == "amdgpu" {
+        if ptr.layout.is_zst() {
+            return OperandRef::new_zst(self.cx(), ptr.layout);
+        }
+
+        fn scalar_load_metadata<'a, 'll, 'tcx>(
+            bx: &mut Builder<'a, 'll, 'tcx, &'ll Value>,
+            load: &'ll Value,
+            scalar: &layout::Scalar
+        ) {
+            let vr = scalar.valid_range.clone();
+            match scalar.value {
+                layout::Int(..) => {
+                    let range = scalar.valid_range_exclusive(bx.cx());
+                    if range.start != range.end {
+                        bx.range_metadata(load, range);
+                    }
+                }
+                layout::Pointer if vr.start() < vr.end() && !vr.contains(&0) => {
+                    bx.nonnull_metadata(load);
+                }
+                _ => {}
+            }
+        }
+
+        let val = if let Some(llextra) = ptr.llextra {
+            OperandValue::Ref(ptr.llval, Some(llextra), ptr.align)
+        } else if ptr.layout.is_llvm_immediate() {
+            let mut const_llval = None;
+            unsafe {
+                if let Some(global) = llvm::LLVMIsAGlobalVariable(ptr.llval) {
+                    if llvm::LLVMIsGlobalConstant(global) == llvm::True {
+                        const_llval = llvm::LLVMGetInitializer(global);
+                    }
+                }
+            }
+            let llval = const_llval.unwrap_or_else(|| {
+                let load = self.load(ptr.llval, ptr.align);
+                if let layout::Abi::Scalar(ref scalar) = ptr.layout.abi {
+                    scalar_load_metadata(self, load, scalar);
+                }
+                load
+            });
+            OperandValue::Immediate(to_immediate(self, llval, ptr.layout))
+        } else if let layout::Abi::ScalarPair(ref a, ref b) = ptr.layout.abi {
+            let mut load = |i, scalar: &layout::Scalar| {
+                let llptr = self.struct_gep(ptr.llval, i as u64);
+                let load = self.load(llptr, ptr.align);
+                scalar_load_metadata(self, load, scalar);
+                if scalar.is_bool() {
+                    self.trunc(load, self.cx().type_i1())
+                } else {
+                    load
+                }
+            };
+            OperandValue::Pair(load(0, a), load(1, b))
+        } else {
+            OperandValue::Ref(ptr.llval, None, ptr.align)
+        };
+
+        OperandRef { val, layout: ptr.layout }
+    }
+
+
+
+    fn range_metadata(&mut self, load: &'ll Value, range: Range<u128>) {
```

Comment:
I don't think we should expose this like this, but maybe as an optional argument to `load`?